File: services/notification-service/internal/consumer/consumer.py
import json
import logging
import uuid

from nats.aio.msg import Msg
from nats.js import JetStreamContext
from nats.js.api import AckPolicy, ConsumerConfig, DeliverPolicy

from internal.broker.connection import SETTLEMENT_EVENTS_STREAM
from internal.notifications.repository import NotificationRepository

logger = logging.getLogger(__name__)

RISK_HOLD_DURABLE = "notification-service-risk-hold"
SETTLEMENT_FINALIZED_DURABLE = "notification-service-settlement-finalized"


def decide_risk_scored(payload: dict) -> dict | None:
    if payload.get("decision") != "hold":
        return None
    return {
        "type": "risk_hold",
        "subject_id": uuid.UUID(payload["transaction_id"]),
        "payload": payload,
    }


def decide_settlement_finalized(payload: dict) -> dict:
    return {
        "type": "settlement_finalized",
        "subject_id": uuid.UUID(payload["settlement_id"]),
        "payload": payload,
    }


class Consumer:
    def __init__(self, repo: NotificationRepository):
        self._repo = repo

    async def start(self, js: JetStreamContext) -> None:
        await js.subscribe(
            "transaction.risk-scored",
            stream=SETTLEMENT_EVENTS_STREAM,
            durable=RISK_HOLD_DURABLE,
            cb=self._handle_risk_scored,
            manual_ack=True,
            config=ConsumerConfig(
                durable_name=RISK_HOLD_DURABLE,
                ack_policy=AckPolicy.EXPLICIT,
                deliver_policy=DeliverPolicy.ALL,
            ),
        )
        await js.subscribe(
            "settlement.finalized",
            stream=SETTLEMENT_EVENTS_STREAM,
            durable=SETTLEMENT_FINALIZED_DURABLE,
            cb=self._handle_settlement_finalized,
            manual_ack=True,
            config=ConsumerConfig(
                durable_name=SETTLEMENT_FINALIZED_DURABLE,
                ack_policy=AckPolicy.EXPLICIT,
                deliver_policy=DeliverPolicy.ALL,
            ),
        )

    async def _handle_risk_scored(self, msg: Msg) -> None:
        await self._handle(msg, decide_risk_scored)

    async def _handle_settlement_finalized(self, msg: Msg) -> None:
        await self._handle(msg, decide_settlement_finalized)

    async def _handle(self, msg: Msg, decide) -> None:
        try:
            payload = json.loads(msg.data)
        except json.JSONDecodeError:
            logger.error("consumer: malformed payload on %s, terminating message", msg.subject)
            await msg.term()
            return

        record = decide(payload)
        if record is not None:
            try:
                self._repo.record(record["type"], record["subject_id"], record["payload"])
            except Exception:
                logger.exception("consumer: failed to record notification for %s", msg.subject)
                await msg.nak()
                return

        await msg.ack()

File: services/settlement-engine/internal/consumer/consumer.go
package consumer

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"time"

	"github.com/nats-io/nats.go/jetstream"

	"github.com/phuoctmse/settleguard/settlement-engine/internal/broker"
	"github.com/phuoctmse/settleguard/settlement-engine/internal/ledgerevent"
	"github.com/phuoctmse/settleguard/settlement-engine/internal/risk"
	"github.com/phuoctmse/settleguard/settlement-engine/internal/settlement"
)

// DurableName is the JetStream durable consumer name settlement-engine
// uses on the LEDGER_EVENTS stream. Changing FilterSubject/AckPolicy for
// an existing durable of this name later requires deleting and
// recreating it server-side -- CreateOrUpdateConsumer reconciles
// compatible changes but not every field is mutable in place.
const DurableName = "settlement-engine-risk-scoring"

const scoreTimeout = 10 * time.Second

// Consumer scores each ledger.entry-recorded event and persists the
// result via TransactionRepository.RecordScore.
type Consumer struct {
	scorer       *risk.Scorer
	transactions *settlement.TransactionRepository
}

func New(scorer *risk.Scorer, transactions *settlement.TransactionRepository) *Consumer {
	return &Consumer{scorer: scorer, transactions: transactions}
}

// Start creates (or attaches to) the durable JetStream consumer and begins
// consuming. The caller must Stop() the returned ConsumeContext on
// shutdown.
func (c *Consumer) Start(ctx context.Context, js jetstream.JetStream) (jetstream.ConsumeContext, error) {
	cons, err := js.CreateOrUpdateConsumer(ctx, broker.LedgerEventsStream, jetstream.ConsumerConfig{
		Durable:       DurableName,
		FilterSubject: ledgerevent.EventLedgerEntryRecorded,
		AckPolicy:     jetstream.AckExplicitPolicy,
		DeliverPolicy: jetstream.DeliverAllPolicy,
	})
	if err != nil {
		return nil, fmt.Errorf("create/attach consumer %s: %w", DurableName, err)
	}

	return cons.Consume(c.handleMessage)
}

func (c *Consumer) handleMessage(msg jetstream.Msg) {
	var payload ledgerevent.OutboxPayload
	if err := json.Unmarshal(msg.Data(), &payload); err != nil {
		log.Printf("consumer: malformed payload, terminating message: %v", err)
		_ = msg.Term()
		return
	}

	accountIDs := ledgerevent.AccountIDs(payload.Entries)
	amount := ledgerevent.TotalAmount(payload.Entries)

	ctx, cancel := context.WithTimeout(context.Background(), scoreTimeout)
	defer cancel()

	score, err := c.scorer.Score(ctx, risk.TransactionInput{
		ID:         payload.TransactionID,
		AccountIDs: accountIDs,
		Amount:     amount,
		OccurredAt: ledgerevent.OccurredAt(payload.Entries),
	})
	if err != nil {
		log.Printf("consumer: score transaction %s: %v", payload.TransactionID, err)
		_ = msg.Nak()
		return
	}

	// A held transaction is a successful, expected outcome of scoring --
	// not an error -- so it still gets Ack()'d here like a passing one.
	if err := c.transactions.RecordScore(ctx, accountIDs, amount, score); err != nil {
		log.Printf("consumer: record score for transaction %s: %v", payload.TransactionID, err)
		_ = msg.Nak()
		return
	}

	if err := msg.Ack(); err != nil {
		log.Printf("consumer: ack transaction %s: %v", payload.TransactionID, err)
	}
}
